app_server.py

```diff
@@ -239,218 +239,108 @@ def heartbeat(self):
-                return (unresolved_identifier.status, message)
-        else:
-            # All metadata lookups have completed. Create that OPDS
-            # entry!
-            entry = identifier.opds_entry()
-
-        if entry is None:
-            # This app can't do lookups on an identifier of this
-            # type, so the best thing to do is to treat this
-            # identifier as a 404 error.
-            return (404, self.UNRECOGNIZED_IDENTIFIER)
-
-        # We made it!
-        return entry
-
-    def work_lookup(self, annotator, route_name='lookup',
-                    require_active_licensepool=True, collection=None):
+    def work_lookup(self, annotator, route_name='lookup', collection=None,
+                    urns=[]):
         """Generate an OPDS feed describing works identified by identifier."""
         urns = flask.request.args.getlist('urn')
 
         this_url = cdn_url_for(route_name, _external=True, urn=urns)
         for urn in urns:
-            self.process_urn(urn, collection=collection)
-
-        # The commit is necessary because we may have registered new
-        # Identifier or UnresolvedIdentifier objects.
-        self._db.commit()
+            self.process_urn(urn, collection)
+        self.post_lookup_hook()
 
         opds_feed = LookupAcquisitionFeed(
             self._db, "Lookup results", this_url, self.works, annotator,
             messages_by_urn=self.messages_by_urn, 
             precomposed_entries=self.precomposed_entries,
-            require_active_licensepool=require_active_licensepool
         )
         return feed_response(opds_feed)
 
-    def permalink(self, urn, annotator):
-        """Generate an OPDS feed for looking up a single work by identifier."""
-        this_url = cdn_url_for('work', _external=True, urn=urn)
+    def permalink(self, urn, annotator, route_name='work'):
+        """Look up a single identifier and generate an OPDS feed."""
+        this_url = cdn_url_for(route_name, _external=True, urn=urn)
         self.process_urn(urn)
+        self.post_lookup_hook()
 
-        # The commit is necessary because we may have registered new
-        # Identifier or UnresolvedIdentifier objects.
-        self._db.commit()
-
+        # A LookupAcquisitionFeed's .works is a list of (identifier,
+        # work) tuples, but an AcquisitionFeed's .works is just a
+        # list of works.
+        works = [work for (identifier, work) in self.works]
         opds_feed = AcquisitionFeed(
-            self._db, urn, this_url, self.works, annotator,
-            messages_by_urn=self.messages_by_urn)
+            self._db, urn, this_url, works, annotator,
+            messages_by_urn=self.messages_by_urn
+        )
 
         return feed_response(opds_feed)
+    
+    def process_urn(self, urn, collection=None):
+        """Turn a URN into a Work suitable for use in an OPDS feed.
 
+        :param collection: Not used in this base class.
```

Comment:
I want to keep it in the method signature because otherwise it causes a big problem in metadata. I'll change this method to take **kwargs instead.